torch/_dynamo/output_graph.py

```diff
@@ -441,27 +476,57 @@ def wrap_name(module_key):
                 )
 
         assert self.nn_modules is not None
-        for k, v in self.nn_modules.items():
-            if v is target:
-                # it already exists
-                return wrap_name(k)
+        # Each param has a source, and even if they have the same tensor value, it needs to
+        # be reachable. So, for ParamBufferSource, which is the leaf node source initialized below
+        # we can skip this value check. It's wrap_name is a no op anyway, so this is safe.
+        if not isinstance(source, ParamBufferSource):
+            for k, v in self.nn_modules.items():
+                if v is target:
+                    # it already exists, but register it anyway
+                    return wrap_name(k)
 
         # create a new unique name
         name = "_".join(map(str, names))
-        # e.g. replace abc.xyz[123].qkv with abc.xyz_123.qkv
-        name = re.sub(r"\[(\d+)\]", r"_\g<1>", name)
-        # e.g. replace abc.xyz_123.qkv with abc_xyz_123_qkv
-        name = re.sub(r"[^a-zA-Z0-9]", "_", name)
 
+        # NOTE! This code used to strip square brackets,
+        # but we no longer do so in favor of a single uniformly used
+        # normalize_attr_name. We used to need this because of things we did with module
+        # lookups where we would eval with a name, and the square brackets would throw it off.
+        # We no longer do this.
+        #
+        # The reason we need this uniform is that we need all the names to be the same, so that
+        # all the sources can be the same, because when we eventually call .register into the
+        # context_module, we need to make sure elements match.
+        name = get_or_make_known_name(name, source)
         if not name or not name[0].isalpha():
             name = "sub" + name
         base = name
         for i in itertools.count():
```

Comment:
I'm going to keep rejecting this PR until this loop no longer exists. ONE spot for name allocation, not two.